parameter_traits/test/validators_tests.cpp

```diff
@@ -0,0 +1,460 @@
+//      notice, this list of conditions and the following disclaimer.
+//
+//    * Redistributions in binary form must reproduce the above copyright
+//      notice, this list of conditions and the following disclaimer in the
+//      documentation and/or other materials provided with the distribution.
+//
+//    * Neither the name of the copyright holder nor the names of its
+//      contributors may be used to endorse or promote products derived from
+//      this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include "gtest/gtest.h"
+
+#include <string>
+#include <vector>
+
+#include <rclcpp/parameter.hpp>
+
+#include <parameter_traits/result.hpp>
+#include <parameter_traits/validators.hpp>
+
+namespace parameter_traits {
+using rclcpp::Parameter;
+
+TEST(ValidatorsTests, Bounds) {
+  EXPECT_TRUE(bounds<double>(Parameter{"", 1.0}, 1.0, 5.0).success());
+  EXPECT_TRUE(bounds<double>(Parameter{"", 4.3}, 1.0, 5.0).success());
+  EXPECT_TRUE(bounds<double>(Parameter{"", 5.0}, 1.0, 5.0).success());
+  EXPECT_FALSE(bounds<double>(Parameter{"", -4.3}, 1.0, 5.0).success());
+  EXPECT_FALSE(bounds<double>(Parameter{"", 10.2}, 1.0, 5.0).success());
+
+  EXPECT_TRUE(bounds<int64_t>(Parameter{"", 1}, 1, 5).success());
+  EXPECT_TRUE(bounds<int64_t>(Parameter{"", 4}, 1, 5).success());
+  EXPECT_TRUE(bounds<int64_t>(Parameter{"", 5}, 1, 5).success());
+  EXPECT_FALSE(bounds<int64_t>(Parameter{"", -4}, 1, 5).success());
+  EXPECT_FALSE(bounds<int64_t>(Parameter{"", 10}, 1, 5).success());
+
+  EXPECT_TRUE(bounds<bool>(Parameter{"", true}, false, true).success());
+  EXPECT_FALSE(bounds<bool>(Parameter{"", false}, true, true).success());
+  EXPECT_FALSE(bounds<bool>(Parameter{"", true}, false, false).success());
+
+  EXPECT_ANY_THROW(bounds<int64_t>(Parameter{"", ""}, 1, 5));
+  EXPECT_ANY_THROW(bounds<std::string>(Parameter{"", 4}, "", "foo"));
+}
+
+TEST(ValidatorsTests, LowerBounds) {
+  EXPECT_TRUE(lower_bounds<double>(Parameter{"", 2.0}, 2.0).success());
+  EXPECT_TRUE(lower_bounds<double>(Parameter{"", 4.3}, 1.0).success());
+  EXPECT_FALSE(lower_bounds<double>(Parameter{"", -4.3}, 1.0).success());
```

Comment:
Why call it lower_bounds instead of something like greater_than_or_equal?

Reply:
I started with a single `bounds` function and these grew out of that. When working through this yesterday I had a similar thought about making these map better for operators.